File: platforms/android/runtime/src/main/kotlin/rs/whisker/runtime/scene/HostEventGate.kt
package rs.whisker.runtime.scene

/** Prevents Host element callbacks from re-entering Rust while a frame is being presented. */
internal class HostEventGate(
    private val schedule: (() -> Unit) -> Unit,
) {
    private var isApplyingFrame = false
    private val deferred = ArrayList<() -> Unit>()

    fun beginFrame() {
        check(!isApplyingFrame)
        isApplyingFrame = true
    }

    fun endFrame() {
        check(isApplyingFrame)
        isApplyingFrame = false
        if (deferred.isEmpty()) return
        val events = deferred.toList()
        deferred.clear()
        schedule {
            events.forEach { it() }
        }
    }

    fun dispatch(event: () -> Unit) {
        if (isApplyingFrame) deferred += event else event()
    }
}

File: platforms/ios/Sources/WhiskerRuntime/scene/HostEventGate.swift
import Foundation

/// Prevents Host element callbacks from re-entering Rust while a frame is being presented.
final class HostEventGate {
    private let schedule: (@escaping () -> Void) -> Void
    private var isApplyingFrame = false
    private var deferred: [() -> Void] = []

    init(schedule: @escaping (@escaping () -> Void) -> Void) {
        self.schedule = schedule
    }

    func beginFrame() {
        precondition(!isApplyingFrame)
        isApplyingFrame = true
    }

    func endFrame() {
        precondition(isApplyingFrame)
        isApplyingFrame = false
        guard !deferred.isEmpty else { return }
        let events = deferred
        deferred.removeAll(keepingCapacity: true)
        schedule {
            events.forEach { $0() }
        }
    }

    func dispatch(_ event: @escaping () -> Void) {
        if isApplyingFrame {
            deferred.append(event)
        } else {
            event()
        }
    }
}
